### ISSUE TITLE
Use System.Runtime.ProfileOptimization for faster startup

### ISSUE THREAD

**Issue by benaadams** [dotnet/aspnetcore#18605]
Default host should use ProfileOptimization ([docs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.profileoptimization?view=netcore-3.1)) for faster startup on after first run.

e.g. it should run something like as follows as early as possible during startup:
```csharp
ProfileOptimization.SetProfileRoot( <existing path> );
ProfileOptimization.StartProfile("startup.pgo");
```

**Comment by AndyAyersMS** [comment]
Should look into whether this eagerly jits at Tier1. If so, you might win at startup but miss out on some optimizations, and so end up with lower steady state perf.

**Comment by benaadams** [comment]
Raised issue https://github.com/dotnet/runtime/issues/2273

**Comment by anurse** [comment]
Should hosting do this for you automatically though? Does it have a negative impact on your app at all?

**Comment by benaadams** [comment]
> Should hosting do this for you automatically though?

Standard for vintage ASP.NET?

> [ASP.NET Performance: Prefetch and Multi-Core Jitting](https://devblogs.microsoft.com/aspnet/asp-net-performance-prefetch-and-multi-core-jitting/)
>
> ### Multi-Core Jitting feature.
>
> The multi-core jitting feature is enabled by default, and work on all OS where ASP.NET 4.5 is available. You can disable it by setting the profileGuidedOptimizations enumerated value on the `<system.web/compilation>` section on the web.config, to `None`.
>
> ### Multi-Core Jitting Implementation.
>
> The feature uses the new class ProfileOptimization class of the System.Runtime namespace. The class is called the first time the app domain is created. First, we set the profile root to the codegen directory, which by default is something like:
>
> `“%windir%Microsoft.NETFramework64v4.0.30319Temporary ASP.NET Filesmyappweirdhash1weirdhash2″`
>
> Remember that the temporary files folder value can be changed on tempDirectory attribute of the `<system.web/compilation>` section of the web.config.
> 
> After setting the profile root, we call `StartProfile()` with the filename `“profileoptimization.prof”`. The CLR code underneath handles all the magic. If you examine the contents of the temporary ASP.NET file folder, you will see the file there:
>    
>    ...
>  ### Conclusion 
>
> ... 
>
> The multi-core jitting feature is available in more operating systems and is on by default. It uses a new CLR feature for profile base optimization.

**Comment by benaadams** [comment]
> Does it have a negative impact on your app at all?

No idea

**Comment by rynowak** [comment]
The hard part is this (which you kinda glossed over 😆 )

```C#
ProfileOptimization.SetProfileRoot( <existing directory> );
``` 

When we've done investigations of MCJ in the past the benefit was *really* minimal in the best case, like 5%. There's a high amount of complexity in us being able to compute a stable directory to use for MCG as well as understanding when it's applicable or worth it. If we add this code (inside hosting) it has to work **everywhere** without any drawbacks.

That combined with the low benefit is why we haven't done this.

@davidfowl @DamianEdwards in case they have extra special thoughts.

**Comment by DamianEdwards** [comment]
IIRC, most of the benefit we saw for the app and environment we were testing with was that it parallelized the file I/O to load assemblies, rather than a huge benefit for the actual JIT itself.

Like @rynowak said, the idea of a universally safe place to write to in order to get the benefit (required to do this in the host) was where we tend to get stuck on this. I agree that supporting it for apps themselves to profile and set is interesting though.

**Comment by rampaa** [comment]
> Should look into whether this eagerly jits at Tier1. If so, you might win at startup but miss out on some optimizations, and so end up with lower steady state perf.

Based on https://github.com/dotnet/runtime/issues/2273, am I correct in understanding that MultiCore JIT does NOT cause us to miss out on optimizations?

**Comment by AndyAyersMS** [comment]
> > Should look into whether this eagerly jits at Tier1. If so, you might win at startup but miss out on some optimizations, and so end up with lower steady state perf.
> 
> Based on [dotnet/runtime#2273](https://github.com/dotnet/runtime/issues/2273), am I correct in understanding that MultiCore JIT does NOT cause us to miss out on optimizations?

Yes, if the MCJ initial jitting really produces Tier0 methods. Should be easy enough to check.

**Comment by rampaa** [comment]
> Yes, if the MCJ initial jitting really produces Tier0 methods. Should be easy enough to check.

I'm not sure how I can do this. I tried to inspect the profile file but the human-readable parts of it doesn't seem to include anything about the JIT tier. Would you point me to right direction so that I can check whether the MCJ initial jitting really produces Tier0 methods?
